### PR TITLE
clean up hero controller scope

### PR DESCRIPTION
**Pull request by chunhtai**
## Description

This pr cleans up the comments in https://github.com/flutter/flutter/pull/58808 and makes the herocontroller swapping more robust

## Related Issues

https://github.com/flutter/flutter/issues/45938

## Tests

I added the following tests:

See files

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by goderbauer** [comment]
Looks like cirrus is unhappy.

**Comment by chunhtai** [comment]
filed a issue for customer to fix the test https://github.com/alibaba/flutter_boost/issues/780

In the meanwhile, this is ready for review

**Comment by chunhtai** [comment]
a friendly bump 👍

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 1]
```diff
@@ -645,6 +640,14 @@ class HeroControllerScope extends InheritedWidget {
     Widget child,
   }) : super(key: key, child: child);
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart:641]: given the .none constructor, should we assert it's non-null here?

**Comment by Hixie** [comment]
seems reasonable, modulo test failures

**Comment by chunhtai** [comment]
@Hixie I have a pr to remove the test. https://github.com/flutter/tests/pull/42

**Comment by chunhtai** [comment]
text passed, ready for review! @goderbauer @Hixie

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**